slice key/value bound check error

Go+ code
```
// a := []int{1:2, 3:1}  //build ok
a := []int{3:1,1:2} // build fail
println(a)
```
```
panic: runtime error: index out of range [3] with length 2
```

--- a/cl/expr.go
+++ b/cl/expr.go
@@ -142,7 +142,12 @@ func compileIdentLHS(ctx *blockCtx, name string, mode compileMode) {
 		if mode == token.ASSIGN || mode == token.DEFINE {
 			ctx.out.StoreVar(v.v)
 		} else if op, ok := addrops[mode]; ok {
-			ctx.out.AddrVar(v.v).AddrOp(kindOf(v.v.Type()), op)
+			typ := v.v.Type()
+			if typ.Kind() == reflect.Ptr {
+				ctx.out.LoadVar(v.v).AddrOp(kindOf(typ), op)
+			} else {
+				ctx.out.AddrVar(v.v).AddrOp(kindOf(typ), op)
+			}
 		} else {
 			log.Panicln("compileIdentLHS failed: unknown op -", mode)
 		}

--- a/cl/expr.go
+++ b/cl/expr.go
@@ -313,7 +313,6 @@ func compileCompositeLit(ctx *blockCtx, v *ast.CompositeLit) func() {
 					elts[n] = e
 				}
 			}
-			n++
 			typElem := typSlice.Elem()
 			for _, elt := range elts {
 				if elt != nil {
@@ -323,7 +322,7 @@ func compileCompositeLit(ctx *blockCtx, v *ast.CompositeLit) func() {
 					ctx.out.Zero(typElem)
 				}
 			}
-			ctx.out.MakeArray(typSlice, n)
+			ctx.out.MakeArray(typSlice, nLen)
 		}
 	case reflect.Map:
 		typMap := typ.(reflect.Type)

--- a/cl/expr.go
+++ b/cl/expr.go
@@ -890,15 +890,15 @@ func compileIndexExprLHS(ctx *blockCtx, v *ast.IndexExpr, mode compileMode) {
 
 	typ := ctx.infer.Get(-1).(iValue).Type()
 	typElem := typ.Elem()
+	if typ.Kind() == reflect.Array {
+		ctx.checkLoadAddr = true
+	}
 	if typ.Kind() == reflect.Ptr {
 		if typElem.Kind() != reflect.Array {
 			logPanic(ctx, v, `type %v does not support indexing`, typ)
 		}
 		typ = typElem
 		typElem = typElem.Elem()
-	}
-	if typ.Kind() == reflect.Array {
-		ctx.checkLoadAddr = true
 	}
 	exprX()
 	ctx.checkLoadAddr = false

--- a/cl/expr.go
+++ b/cl/expr.go
@@ -300,12 +300,7 @@ func compileCompositeLit(ctx *blockCtx, v *ast.CompositeLit) func() {
 		}
 		ctx.infer.Push(&goValue{t: typSlice})
 		return func() {
-			var nLen int
-			if kind == reflect.Array {
-				nLen = typSlice.Len()
-			} else {
-				nLen = toBoundArrayLen(ctx, v)
-			}
+			nLen := toBoundArrayLen(ctx, v)
 			n := -1
 			elts := make([]ast.Expr, nLen)
 			for _, elt := range v.Elts {

--- a/cl/expr_test.go
+++ b/cl/expr_test.go
@@ -533,11 +533,10 @@ func TestArray(t *testing.T) {
 	cltest.Expect(t, `
 		x := [4]float64{1, 2.3, 3.6}
 		println("x:", x)
-
 		y := [...]float64{1, 2.3, 3.6}
 		println("y:", y)
 		`,
-		"x: [2.3 3.6 0 0]\ny: [1 2.3 3.6]\n",
+		"x: [1 2.3 3.6 0]\ny: [1 2.3 3.6]\n",
 	)
 	cltest.Expect(t, `
 		x := [...]float64{1, 3: 3.4, 5}

--- a/cl/expr_test.go
+++ b/cl/expr_test.go
@@ -527,6 +527,14 @@ func TestSlice(t *testing.T) {
 		`,
 		"x: [1 0 3.4 5]\n",
 	)
+	cltest.Expect(t, `
+		x := []float64{1:1,3:3,4}
+		println("x:", x)
+		y := []float64{3:3,4,1:1}
+		println("y:", x)
+		`,
+		"x: [0 1 0 3 4]\ny: [0 1 0 3 4]\n",
+	)
 }
 
 func TestArray(t *testing.T) {
@@ -537,6 +545,14 @@ func TestArray(t *testing.T) {
 		println("y:", y)
 		`,
 		"x: [1 2.3 3.6 0]\ny: [1 2.3 3.6]\n",
+	)
+	cltest.Expect(t, `
+		x := [5]float64{1:1,3:3,4}
+		println("x:", x)
+		y := [5]float64{3:3,4,1:1}
+		println("y:", x)
+		`,
+		"x: [0 1 0 3 4]\ny: [0 1 0 3 4]\n",
 	)
 	cltest.Expect(t, `
 		x := [...]float64{1, 3: 3.4, 5}

--- a/cl/type_decl.go
+++ b/cl/type_decl.go
@@ -289,12 +289,20 @@ func toArrayType(ctx *blockCtx, v *ast.ArrayType) iType {
 
 func toBoundArrayLen(ctx *blockCtx, v *ast.CompositeLit) int {
 	n := -1
+	var max int
 	for _, elt := range v.Elts {
 		if e, ok := elt.(*ast.KeyValueExpr); ok {
-			n = toInt(ctx, e.Key)
+			i := toInt(ctx, e.Key)
+			if i < n {
+				max = n
+			}
+			n = i
 		} else {
 			n++
 		}
+	}
+	if n < max {
+		n = max
 	}
 	return n + 1
 }

--- a/cl/type_decl.go
+++ b/cl/type_decl.go
@@ -289,7 +289,7 @@ func toArrayType(ctx *blockCtx, v *ast.ArrayType) iType {
 
 func toBoundArrayLen(ctx *blockCtx, v *ast.CompositeLit) int {
 	n := -1
-	var max int
+	max := -1
 	for _, elt := range v.Elts {
 		if e, ok := elt.(*ast.KeyValueExpr); ok {
 			i := toInt(ctx, e.Key)

--- a/exec/bytecode/var.go
+++ b/exec/bytecode/var.go
@@ -382,9 +382,6 @@ func (p *Builder) StoreVar(v *Var) *Builder {
 
 // AddrVar instr
 func (p *Builder) AddrVar(v *Var) *Builder {
-	if v.Type().Kind() == reflect.Ptr {
-		return p.LoadVar(v)
-	}
 	p.addrVar(makeAddr(p.nestDepth-v.nestDepth, v.idx))
 	return p
 }

--- a/exec/golang/var.go
+++ b/exec/golang/var.go
@@ -224,6 +224,9 @@ func (p *Builder) bigAddrOp(kind exec.Kind, op exec.AddrOperator) *Builder {
 		}
 		bigOp := &ast.SelectorExpr{X: v.X, Sel: Ident(method)}
 		expr = &ast.CallExpr{Fun: bigOp, Args: []ast.Expr{v.X, val}}
+	case *ast.Ident:
+		bigOp := &ast.SelectorExpr{X: v, Sel: Ident(method)}
+		expr = &ast.CallExpr{Fun: bigOp, Args: []ast.Expr{v, val}}
 	default:
 		log.Panicln("bigAddrOp: todo")
 	}

